### PR TITLE
mgr/volumes: add `mypy` support

### PR DESCRIPTION
**Pull request by mgfritch**
Adds `mypy` checks to the `mgr/volumes` modules.

Fixes: https://tracker.ceph.com/issues/44393
Signed-off-by: Michael Fritch <mfritch@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mgfritch** [comment]
This caught a large number of legitimate bugs!

```
mypy run-test-pre: PYTHONHASHSEED='304868522'
mypy run-test: commands[0] | mypy --config-file=../../mypy.ini cephadm/module.py cephadm/_utils.py mgr_module.py dashboard/module.py mgr_util.py orchestrator/__init__.py progress/module.py rook/module.py osd_support/module.py test_orchestrator/module.py volumes/__init__.py
volumes/fs/async_cloner.py:13: note: In module imported here,
volumes/fs/volume.py:18: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/op_sm.py: note: In member "get_init_state" of class "OpSm":
volumes/fs/operations/op_sm.py:36: error: Name 'errno' is not defined
volumes/fs/operations/op_sm.py:37: error: "object" has no attribute "get"
volumes/fs/operations/op_sm.py:39: error: Name 'errno' is not defined
volumes/fs/operations/op_sm.py: note: In member "get_next_state" of class "OpSm":
volumes/fs/operations/op_sm.py:46: error: Name 'errno' is not defined
volumes/fs/operations/op_sm.py:47: error: "object" has no attribute "get"
volumes/fs/operations/op_sm.py:49: error: Name 'errno' is not defined
volumes/fs/operations/group.py:9: note: In module imported here,
volumes/fs/volume.py:12: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/template.py: note: In member "list_subvolumes" of class "GroupTemplate":
volumes/fs/operations/template.py:5: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "create_snapshot" of class "GroupTemplate":
volumes/fs/operations/template.py:14: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "remove_snapshot" of class "GroupTemplate":
volumes/fs/operations/template.py:23: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "list_snapshots" of class "GroupTemplate":
volumes/fs/operations/template.py:32: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "open" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:42: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "status" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:45: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "create" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:62: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "create_clone" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:74: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "remove" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:85: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "resize" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:94: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "create_snapshot" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:103: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "remove_snapshot" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:112: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "snapshot_path" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:121: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "list_snapshots" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:130: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "is_snapshot_protected" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:139: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "protect_snapshot" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:148: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "unprotect_snapshot" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:158: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "attach_snapshot" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:169: error: Name 'errno' is not defined
volumes/fs/operations/template.py: note: In member "detach_snapshot" of class "SubvolumeTemplate":
volumes/fs/operations/template.py:180: error: Name 'errno' is not defined
volumes/fs/operations/volume.py:17: note: In module imported here,
volumes/fs/volume.py:10: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/lock.py: note: In member "__init__" of class "GlobalLock":
volumes/fs/operations/lock.py:27: error: "object" has no attribute "__enter__"
volumes/fs/operations/lock.py:27: error: "object" has no attribute "__exit__"
volumes/fs/operations/lock.py: note: In member "lock_op" of class "GlobalLock":
volumes/fs/operations/lock.py:35: error: "object" has no attribute "__enter__"
volumes/fs/operations/lock.py:35: error: "object" has no attribute "__exit__"
volumes/fs/operations/versions/subvolume_base.py:9: note: In module imported here,
volumes/fs/operations/versions/__init__.py:7: note: ... from here,
volumes/fs/operations/subvolume.py:11: note: ... from here,
volumes/fs/volume.py:13: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/versions/metadata_manager.py:8: error: Name 'configparser' already defined (by an import)
volumes/fs/operations/clone_index.py:10: note: In module imported here,
volumes/fs/async_cloner.py:18: note: ... from here,
volumes/fs/volume.py:18: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/index.py: note: In member "track" of class "Index":
volumes/fs/operations/index.py:18: error: Name 'VolumeException' is not defined
volumes/fs/operations/index.py:18: error: Name 'errno' is not defined
volumes/fs/operations/index.py: note: In member "untrack" of class "Index":
volumes/fs/operations/index.py:21: error: Name 'VolumeException' is not defined
volumes/fs/operations/index.py:21: error: Name 'errno' is not defined
volumes/fs/async_cloner.py:10: note: In module imported here,
volumes/fs/volume.py:18: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/async_job.py: note: In member "run" of class "JobThread":
volumes/fs/async_job.py:43: error: "Thread" has no attribute "should_cancel"
volumes/fs/async_job.py: note: In member "should_cancel" of class "JobThread":
volumes/fs/async_job.py:69: error: "Event" has no attribute "isSet"
volumes/fs/async_job.py: note: In member "__init__" of class "AsyncJobs":
volumes/fs/async_job.py:96: error: Need type annotation for 'q'
volumes/fs/volume.py:8: note: In module imported here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/fs_util.py: note: In function "get_ancestor_xattr":
volumes/fs/fs_util.py:129: error: Too few arguments for "VolumeException"
volumes/fs/async_cloner.py:18: note: In module imported here,
volumes/fs/volume.py:18: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/clone_index.py: note: In member "track" of class "CloneIndex":
volumes/fs/operations/clone_index.py:38: error: Argument 2 to "isinstance" has incompatible type "Union[VolumeException, Any]"; expected "Union[type, Tuple[Union[type, Tuple[Any, ...]], ...]]"
volumes/fs/volume.py:12: note: In module imported here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/group.py: note: In class "Group":
volumes/fs/operations/group.py:44: error: Name 'uid' already defined on line 36
volumes/fs/operations/group.py:44: error: "Callable[[Group], Any]" has no attribute "setter"
volumes/fs/operations/group.py:48: error: Name 'gid' already defined on line 40
volumes/fs/operations/group.py:48: error: "Callable[[Group], Any]" has no attribute "setter"
volumes/fs/operations/group.py: note: In function "create_group":
volumes/fs/operations/group.py:135: error: Not all arguments converted during string formatting
volumes/fs/operations/group.py: note: In function "open_group":
volumes/fs/operations/group.py:170: error: Property "uid" defined in "Group" is read-only
volumes/fs/operations/group.py:171: error: Property "gid" defined in "Group" is read-only
volumes/fs/volume.py:10: note: In module imported here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/volume.py:9: error: Module 'threading' has no attribute '_Timer'; maybe "Timer"?
volumes/fs/operations/volume.py: note: In member "disconnect" of class "Connection":
volumes/fs/operations/volume.py:93: error: Item "None" of "Optional[Any]" has no attribute "get_addrs"
volumes/fs/operations/volume.py:94: error: Item "None" of "Optional[Any]" has no attribute "shutdown"
volumes/fs/operations/volume.py: note: In member "abort" of class "Connection":
volumes/fs/operations/volume.py:104: error: Item "None" of "Optional[Any]" has no attribute "abort_conn"
volumes/fs/operations/volume.py: note: In function "create_volume":
volumes/fs/operations/volume.py:212: error: Too few arguments for "remove_pool"
volumes/fs/operations/volume.py:219: error: Too few arguments for "remove_pool"
volumes/fs/operations/volume.py:220: error: Too few arguments for "remove_pool"
volumes/fs/operations/versions/__init__.py:7: note: In module imported here,
volumes/fs/operations/subvolume.py:11: note: ... from here,
volumes/fs/volume.py:13: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/versions/subvolume_base.py: note: In class "SubvolumeBase":
volumes/fs/operations/versions/subvolume_base.py:45: error: Name 'uid' already defined on line 33
volumes/fs/operations/versions/subvolume_base.py:45: error: "Callable[[SubvolumeBase], Any]" has no attribute "setter"
volumes/fs/operations/versions/subvolume_base.py:49: error: Name 'gid' already defined on line 37
volumes/fs/operations/versions/subvolume_base.py:49: error: "Callable[[SubvolumeBase], Any]" has no attribute "setter"
volumes/fs/operations/versions/subvolume_base.py:53: error: Name 'mode' already defined on line 41
volumes/fs/operations/versions/subvolume_base.py:53: error: "Callable[[SubvolumeBase], Any]" has no attribute "setter"
volumes/fs/volume.py:19: note: In module imported here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/purge_queue.py: note: In function "get_trash_entry_for_volume":
volumes/fs/purge_queue.py:26: error: Cannot find replacement for positional format specifier 1
volumes/fs/purge_queue.py:28: error: Name 'ret' is not defined
volumes/fs/operations/subvolume.py:11: note: In module imported here,
volumes/fs/volume.py:13: note: ... from here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/versions/__init__.py: note: In member "_load_supported_versions" of class "SubvolumeLoader":
volumes/fs/operations/versions/__init__.py:35: error: Too few arguments for "VolumeException"
volumes/fs/operations/versions/__init__.py: note: In member "upgrade_legacy_subvolume" of class "SubvolumeLoader":
volumes/fs/operations/versions/__init__.py:56: error: Name 'OpSmException' is not defined
volumes/fs/operations/versions/__init__.py: note: In member "get_subvolume_object" of class "SubvolumeLoader":
volumes/fs/operations/versions/__init__.py:70: error: Incompatible types in assignment (expression has type "None", variable has type "SubvolumeBase")
volumes/fs/volume.py:18: note: In module imported here,
volumes/module.py:7: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/async_cloner.py: note: In function "get_next_clone_entry":
volumes/fs/async_cloner.py:37: error: Cannot find replacement for positional format specifier 1
volumes/fs/async_cloner.py:39: error: Name 'ret' is not defined
volumes/fs/async_cloner.py: note: In function "handle_clone_pending":
volumes/fs/async_cloner.py:72: error: "OpSmException" has no attribute "error"; maybe "errno"?
volumes/fs/async_cloner.py: note: In function "handle_clone_in_progress":
volumes/fs/async_cloner.py:140: error: "OpSmException" has no attribute "error"; maybe "errno"?
volumes/module.py:7: note: In module imported here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/volume.py: note: In member "is_stopping" of class "VolumeClient":
volumes/fs/volume.py:59: error: "Event" has no attribute "isSet"
volumes/fs/volume.py: note: In member "list_fs_volumes" of class "VolumeClient":
volumes/fs/volume.py:106: error: "Event" has no attribute "isSet"
Found 70 errors in 15 files (checked 11 source files)
```

**Comment by mgfritch** [comment]
Getting these checks working as a pre-req will make life easier if/when we re-use `mgr/volumes` logic from within `mgr/cephadm`.

For example: PR #32297 and PR #33581

**Comment by mgfritch** [comment]
jenkins test make check

**Comment by vshankar** [comment]
Thanks for the effort @mgfritch -- most of these issues (missing imports, typos) are not caught in testing since those are used during exception handling. Nice to have this run w/ mypy.

File: src/pybind/mgr/volumes/fs/operations/index.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import errno
 import os
 
+from .exception import VolumeException
```
[the remark this answers is not in the document]
Comment (mgfritch) [reply src/pybind/mgr/volumes/fs/operations/index.py]: thanks, fixed!

**Comment by mgfritch** [comment]
> This needs a tracker ticker. @mgfritch please create one since we also need to backport this to nautilus.

Added tracker ticket:
https://tracker.ceph.com/issues/44393

**Comment by mgfritch** [comment]
jenkins test make check

**Comment by batrick** [comment]
Conflicts with #33687

**Comment by mgfritch** [comment]
> Conflicts with #33687

dropped the conflicting commit in favor of #33687

**Comment by ajarr** [comment]
Thanks a lot, @mgfritch! We'll need your help backporting this to nautilus too

**Comment by liewegas** [comment]
needs rebase to fix tox error in mgr/volumes

**Comment by mgfritch** [comment]
rebased

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2020-03-06_03:14:09-fs-wip-pdonnell-testing-20200305.225908-distro-basic-smithi/

**Comment by batrick** [comment]
Failure is: https://tracker.ceph.com/issues/44438

**Pull request merged**